mypy/meet.py

```diff
@@ -140,15 +149,28 @@ def visit_void(self, t: Void) -> Type:
             return ErrorType()
 
     def visit_none_type(self, t: NoneTyp) -> Type:
-        if not isinstance(self.s, Void) and not isinstance(self.s, ErrorType):
-            return t
+        if experiments.STRICT_OPTIONAL:
+            if isinstance(self.s, NoneTyp) or (isinstance(self.s, Instance) and
+                                               self.s.type.fullname() == 'builtins.object'):
+                return t
+            else:
+                return UninhabitedType()
         else:
-            return ErrorType()
+            if not isinstance(self.s, Void) and not isinstance(self.s, ErrorType):
+                return t
+            else:
+                return ErrorType()
+
+    def visit_uninhabited_type(self, t: UninhabitedType) -> Type:
+        return t
```

Comment:
Probably meet of `Void` and `UninhabitedType` should still be `ErrorType` I think?